### PR TITLE
Add a property check, so that when the property flyway.migrateOnInit …

### PR DESCRIPTION
**Pull request by ptahchiev**
…is false the flyway migration does not happen.

### CODE REVIEW

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java [hunk 1]
```diff
@@ -25,10 +28,18 @@
  * @author Andreas Ahlenstorf
  * @author Phillip Webb
  */
-public class FlywayMigrationStrategy {
+public class FlywayMigrationStrategy implements EnvironmentAware {
 
-	public void migrate(Flyway flyway) {
-		flyway.migrate();
-	}
+    private Environment environment;
```
Comment (dsyer) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java]: I don't like giving a convenience class full access to the `Environment` when all it needs is a flag. Why not injkect `FlywayProperties`?

Comment (ptahchiev) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java]: OK, I wasn't aware of the `FlywayProperties`. Shall I recreate it?

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java [hunk 2]
```diff
@@ -16,7 +16,10 @@
 
 package org.springframework.boot.autoconfigure.flyway;
 
+import org.apache.commons.lang.BooleanUtils;
```
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java]: Please don't use commons lang

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java (hunk 1 above)
Comment (dsyer) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java]: https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayProperties.java

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java (hunk 2 above)
Comment (dsyer) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayMigrationStrategy.java]: We should add a build rule (we did it for guava right, or was that Spring Cloud?).

**Comment by philwebb** [comment]
@ptahchiev Can you provide some more background about why you want to add a property? Isn't Flyway without migration somewhat useless?

**Comment by ptahchiev** [comment]
I think it is usefuly if you want to allow the customer to manually migrate, and not migrate every time the bean is instantiated. If you call `migrate` method every time the bean is instantiated, then the `flyway-maven-plugin` becomes obsolete - no one will ever use `flyway:migrate`, simply because you have already migrated.

In my case I have a custom maven plugin that calls hibernate's schema export to create the files that flyway needs. Now the customer runs the export `mvn nemesis:exportschema` and because the spring context is loaded and the `Flyway` bean is instantiated the `migrate` method is being called so the database gets populated with database tables, every single time, even though the maven plugin should only *export* the schema.

**Comment by philwebb** [comment]
OK, That makes sense. Thanks for the background info.

**Comment by philwebb** [comment]
I assume that the current `flyway.enabled` flag won't work for you? You need a Flyway bean in your application context, just not migrated? I think you might already be able to do this with 1.3 by implementing your own no-op `FlywayMigrationStrategy`.

**Comment by ptahchiev** [comment]
@philwebb Yep, I need a `Flyway` bean, just not migrated, and yes it all works great with 1.3.0.SNAPSHOT. My suggestion is more a feature, then a real bug.

**Comment by dsyer** [comment]
I still don't understand the motivation for this (or the `FlywayMigrationStrategy`) by the way. I don't think it's harmful to add such a thing, but I don't know why anyone needs it (and I suppose there is some harm in adding features that are not needed.) In your example above, for instance, why do you need a `Flyway` bean at all (if the user is going to populate the database some other way)?

**Comment by ptahchiev** [comment]
@dsyer No, the user is populating the database with the `Flyway` bean. However, I don't want him to do it **every** time the application context initializes the flyway bean. I have a maven plugin that the user manually calls `mvn flyway:migrate`

**Comment by dsyer** [comment]
If you using a build plugin to do the migration, why do you even need flyway on the classpath at runtime?

**Comment by ptahchiev** [comment]
@dsyer because I also offer a much bigger step called `initialization`:
 - migrate the db with flyway.
 - import data using csv files.
 - perform catalog synchronization.
 - export data to solr search server.

For that reason I have also my own maven plugin which uses the flyway API.

**Comment by dsyer** [comment]
I still don't get it. It sounds like that stuff is better done in tooling or as a separate app. Why does your app need Flyway on the classpath?

**Comment by ptahchiev** [comment]
@dsyer well maybe there's a better way to implement it but that's how i've chose to do it. I need Flyway on the classpath because the user can also call a rest endpoint which will take down the website for about 2 minutes to do the initialization process I described above. I also have flyway in the maven plugin, where I also construct an application context.

**Comment by philwebb** [comment]
I think this is probably quite an unusual scenario so it's probably best if we don't add the property. You should be able to implement your own `FlywayMigrationStrategy` to solve the problem.

**Pull request closed without merging**